core/src/main/java/org/apache/calcite/util/ArrowSet.java

```diff
@@ -0,0 +1,357 @@
+        continue;
+      }
+      visited.add(keys);
+      if (onlyMinimalKeys && keys.cardinality() > minKeySize) {
+        break;
+      }
+      boolean covered = false;
+      for (ImmutableBitSet key : result) {
+        if (keys.contains(key)) {
+          covered = true;
+          break;
+        }
+      }
+      if (covered) {
+        continue;
+      }
+      ImmutableBitSet keysClosure = computeClosure(keys);
+      if (keysClosure.contains(ordinals)) {
+        result.add(keys);
+        if (onlyMinimalKeys) {
+          minKeySize = keys.cardinality();
+        }
+        continue;
+      }
+      ImmutableBitSet remain = ordinals.except(keys);
+      for (int attr : remain) {
+        if (keysClosure.get(attr)) {
+          continue;
+        }
+        ImmutableBitSet next = keys.union(ImmutableBitSet.of(attr));
+        if (!visited.contains(next)) {
+          queue.add(next);
+        }
+      }
+    }
+    return result.isEmpty() ? ImmutableSet.of(ordinals) : result;
+  }
+
+  /**
+   * Find ordinals in the given set that never appear as dependents in any functional dependency.
+   * These are the "source" ordinals that cannot be derived from others.
+   */
+  private ImmutableBitSet findNonDependentAttributes(ImmutableBitSet ordinals) {
+    ImmutableBitSet dependentsAttrs = dependencyGraph.values().stream()
+        .reduce(ImmutableBitSet.of(), ImmutableBitSet::union);
+    return ordinals.except(dependentsAttrs);
+  }
+
+  /**
+   * Returns a new ArrowSet that is the union of this and another ArrowSet.
+   */
+  public ArrowSet union(ArrowSet other) {
+    Set<Arrow> unionSet = new HashSet<>();
+    unionSet.addAll(this.getArrows());
+    unionSet.addAll(other.getArrows());
+    return new ArrowSet(unionSet);
+  }
+
+  /**
+   * Returns all arrows (functional dependencies) in this ArrowSet.
+   *
+   * @return a set of Arrow objects representing all functional dependencies
+   */
+  public Set<Arrow> getArrows() {
+    return arrowSet;
+  }
+
+  @Override public ArrowSet clone() {
+    return new ArrowSet(new HashSet<>(this.arrowSet));
+  }
+
+  public boolean equalTo(ArrowSet other) {
+    for (Map.Entry<ImmutableBitSet, ImmutableBitSet> entry : dependencyGraph.entrySet()) {
+      if (!other.implies(entry.getKey(), entry.getValue())) {
+        return false;
+      }
+    }
+    for (Map.Entry<ImmutableBitSet, ImmutableBitSet> entry : other.dependencyGraph.entrySet()) {
+      if (!implies(entry.getKey(), entry.getValue())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("ArrowSet{");
+    boolean first = true;
+    for (Arrow arrow : getArrows()) {
+      if (!first) {
+        sb.append(", ");
+      }
+      sb.append(arrow);
+      first = false;
+    }
+    sb.append('}');
+    return sb.toString();
+  }
+
+  /**
+   * Returns true if this ArrowSet implies that determinants determine dependents.
```

Comment:
if from this ArrowSet one can deduce that 'determinants' determine 'dependents'.
This is a proper use of 'this ArrowSet'.